### DISCUSSION TITLE
Separate ECoS functions to enable different controllers

### DISCUSSION THREAD

**Discussion by big12boy** [cbries/railwayessential#110] [General]
Hey, I've had a quick look at the code and did not see an easy way to add a new controller, like I wanted to.

Have you thought about creating a intermediate class (HAL), which would handle the calls to different controllers?

The config could be extended, like adding a "Controller" key and instead of "Ecos" have something like "Controller" with the values "Ip", "Port" and "ComPort". This would cover pretty much all controllers.

I really like the way your are going with Railessentials, especially as there is no other open source solution which does automation well (I will not go into the Rocrail mess here).

My goal is to add BiDiB (OpenDCC, GBMMaster) support to Railessentials.

Greetings from Austria ;)
- Marcel

**Comment by cbries** [comment]
Hey Marcel.
Good point, at the beginning I thought about that but decided to avoid that step because of missing test  equipment. In addition an abstraction layer of the commands between the ui and command station would be a "nice to have".

I will check the effort for your wish.... should not be very complicated. I will give you a feedback as soon as possible. 

By the way, the dev of the software is paused during summertime. I have to again in Winter!

**Comment by cbries** [comment]
Unfortunatelly it is not that easy as it sounds.
I have to extract several interfaces, e.g. `IDataProvider`, `ILogger`, `IWsServer`, `IItem`, `IDataProvider`, etc. which are weaved into the main core. I do not really know the item/object handling of `BiDiB`, therefor I can not really examine the effort and the correct way to provide an nice HAL/API/Bridge... Currently, I do not have that time to abstract this, the effort -- as I guess -- would be up to 4-5 days (fulltime).

I would appreciate if you try it anyway.... maybe you can just change everything in `ecoslibNet48` to fit your needs, rename all  classes/namespaces and I could add a plugin system...

**Reply by big12boy** [reply]
Ohh ok, yeah I though it was more difficult than you said in the first post, after looking through the code myself.

I will have a look on the weekend, maybe I can try the approach you mentioned.

Thanks anyway for having a look ;)
